torch/distributed/_shard/checkpoint/resharding.py

```diff
@@ -102,140 +81,104 @@ def _shards_get_overlap_region_wrt_saved_tensor(
-    Returns:
-        A 3-element tuple with the following values:
-            List of ``TensorWriteRequest`` for the tensor
-            Metadada describing the tensor.
-            Dictionary describing storage information for this tensor
-
-    NB `storage_key` is used to compose the key names of the local shards.
-    """
-    write_requests = []
-    shard_to_storage_key: Dict[str, str] = dict()
-    storage_md = {}
-
-    for shard_md in sharded_tensor.metadata().shards_metadata:
-        shard_storage_key = _get_shard_storage_key(storage_key, shard_md, storage_key_to_fqn)
-        shard_to_storage_key[_get_shard_key(shard_md)] = shard_storage_key
-        storage_md[MetadataIndex(fqn, shard_md.shard_offsets)] = shard_storage_key
-
-    for shard in sharded_tensor.local_shards():
-        tensor = shard.tensor.detach()
-        shard_storage_key = shard_to_storage_key[_get_shard_key(shard.metadata)]
-
-        wr = TensorWriteRequest(
-            tensor=_trim(tensor),
-            storage_key=shard_storage_key,
-        )
-        write_requests.append(wr)
-    return write_requests, _compute_sharded_tensor_md(sharded_tensor), storage_md
+def _create_read_item_for_byteio(dest_index, dest_offset, storage_index, storage_offset, length):
+    return ReadItem(
+        type=LoadItemType.BYTE_IO,
+        dest_index=dest_index,
+        dest_offsets=torch.Size((dest_offset,)),
+        storage_index=storage_index,
+        storage_offsets=torch.Size((storage_offset,)),
+        lengths=torch.Size((length,)),
+    )
 
 
-def _prepare_sharded_tensor_read(
-    fqn: str,
-    storage_metadata: Dict[MetadataIndex, str],
-    metadata: TensorStorageMetadata,
-    sharded_tensor_out: ShardedTensor
-) -> List[TensorReadRequest]:
-    """
-    Prepare sharded tensor read.
-
-    Args:
-        fqn: The FQN of ``sharded_tensor`` in the state_dict.
-        storage_metadata: Dictionary describing checkpoint storage.
-        metadata: Metadata describing the persisted sharded tensor. Normally,
-                  this is generated by func::`_prepare_sharded_tensor_write`.
-        sharded_tensor_out: The ShardedTensor being read.
-
-    Returns:
-        A list of class::`TensorReadRequest`. When fullfilled,
-        `sharded_tensor_out`'s local shards load from the persisted sharded
-        tensor.
-    """
-    return _prepare_generic_tensor_read(
-        fqn,
-        metadata.chunks,
-        sharded_tensor_out.local_shards(),
-        storage_metadata)
+def _create_read_item_for_tensor(dest_index, dest_offsets, storage_index, storage_offsets, lengths):
+    return ReadItem(
+        type=LoadItemType.TENSOR,
+        dest_index=dest_index,
+        dest_offsets=torch.Size(dest_offsets),
+        storage_index=storage_index,
+        storage_offsets=torch.Size(storage_offsets),
+        lengths=torch.Size(lengths),
+    )
 
-def _prepare_generic_tensor_read(
+def _create_sharded_read_items(
```

Comment:
N00b question, does this mean we now support read from different layout? (Like we save to 10 hosts but when we read, there are only 9 hosts)

Reply:
Yup, we support reading from different distributed layouts, or no distributed at all.
We can go from 10 hosts to 9 hosts to 1 host and no sharded tensor at all.